### PR TITLE
Allow to use a whole local directory as IaC source

### PR DESCRIPTION
**Pull request by eliecharra**
| Q                 | A
| ----------------- | ---
| 🐛 Bug fix?       | no
| 🚀 New feature?   | yes
| ⚠ Deprecations?   | no
| ❌ BC Break       | no
| 🔗 Related issues | #253 
| ❓ Documentation  | yes

## Description

Allow to use a whole local directory as IaC source.
⚠️ Does not follow symlinks within root folder

### CODE REVIEW

**Review by lotoussa — COMMENTED** [review]
Seems cool 👍

**Review by moadibfr — DISMISSED** [review]

**Review by wbeuil — CHANGES_REQUESTED** [review]
Need to change all `multiples` to `multiple`.

I tested it like you did which works perfectly. But we're gonna have problems with people who would like to use it with a folder like this one below which seems pretty normal in terragrunt environment if I remember well:

```
├── route53
│   └── terraform.tf
│   └── terraform.tfstate
└── s3
    └── terraform.tf
    └── terraform.tfstate
```

`driftctl scan` will output:

image

Maybe in another PR we should check that files are indeed terraform.states files before adding them to the slice.

**Comment by eliecharra** [comment]
> Need to change all `multiples` to `multiple`.

Sure nice catch 🙏🏻 

> I tested it like you did which works perfectly. But we're gonna have problems with people who would like to use it with a folder like this one below which seems pretty normal in terragrunt environment if I remember well:

This is a real question, it can totally be done in this PR but I don't know exactly what we want to do. 
The `.tfstate` extension is not enforced by terraform at all, so I don't think we can hardcode a check on this extension.
Maybe it's the responsibility to the user to specify the pattern in the `--from` argument with a glob pattern for example.

I had in mind something like [THAT](https://globster.xyz/?q=**%2F*.tfstate&f=%2Froute53%2Fterraform.tf%2C%2Froute53%2Fterraform.tfstate%2C%2Fs3%2Fterraform.tf%2C%2Fs3%2Fterraform.tfstate%2C%2Fec2%2Fterraform.tf%2C%2Fec2%2Fsubfolder%2Fterraform.tf%2C%2Fec2%2Fsubfolder%2Fterraform.tfstate)

File: pkg/iac/terraform/state/enumerator/file.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+func NewFileEnumerator(config config.SupplierConfig) *FileEnumerator {
+	return &FileEnumerator{
+		config,
+	}
+}
+
+// File enumeration does not follow symlinks.
+// We may use something like this https://pkg.go.dev/github.com/facebookgo/symwalk
+// to follow links, but it allow infinite loop so be careful !
```
[the remark this answers is not in the document]
Comment (eliecharra) [reply pkg/iac/terraform/state/enumerator/file.go:29]: Sounds like I need to take English lesson on the plural rule with you 😅

**Comment by wbeuil** [comment]
> This is a real question, it can totally be done in this PR but I don't know exactly what we want to do.
The .tfstate extension is not enforced by terraform at all, so I don't think we can hardcode a check on this extension.
Maybe it's the responsibility to the user to specify the pattern in the --from argument with a glob pattern for example.

Agree we must not hardcode a check on any extension. Either it's gonna be the user who would specify a pattern like you did or if it's possible we could check if a file is a terraform state but I can't find a method in their repo that does that yet ...

I'd say we could first ship this PR with the changes I requested, then ask @sjourdan about his view and finally decide which way to go.

**Comment by lotoussa** [comment]
> > This is a real question, it can totally be done in this PR but I don't know exactly what we want to do.
> > The .tfstate extension is not enforced by terraform at all, so I don't think we can hardcode a check on this extension.
> > Maybe it's the responsibility to the user to specify the pattern in the --from argument with a glob pattern for example.
> 
> Agree we must not hardcode a check on any extension. Either it's gonna be the user who would specify a pattern like you did or if it's possible we could check if a file is a terraform state but I can't find a method in their repo that does that yet ...
> 
> I'd say we could first ship this PR with the changes I requested, then ask @sjourdan about his view and finally decide which way to go.

Can't we use magic number to check ? I'm not sure there is an identification for tf state tho.

**Comment by moadibfr** [comment]
The scope of ticket in this release was just to be able to use a batch of tfstate inside a repository. We talked about blob filtering but this was more intended as a first step from what I remember.

**Review by wbeuil — APPROVED** [review]
LGTM now

**Pull request merged**